llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -551,102 +686,98 @@ TableFiles processOneModule(std::unique_ptr<Module> M, bool IsEsimd,
   // and these declarations cause an assertion in llvm-spirv. To workaround this
   // issue remove "llvm.used" from the input module before performing any other
   // actions.
-  bool IsLLVMUsedRemoved = false;
-  if (GlobalVariable *GV = M->getGlobalVariable("llvm.used"))
-    IsLLVMUsedRemoved = removeSYCLKernelsConstRefArray(GV);
-
-  if (IsEsimd && LowerEsimd)
-    lowerEsimdConstructs(*M);
-
-  module_split::IRSplitMode Mode = (SplitMode.getNumOccurrences() > 0)
-                                       ? SplitMode
-                                       : module_split::SPLIT_NONE;
-  std::unique_ptr<module_split::ModuleSplitterBase> MSplit =
-      module_split::getSplitterByMode(std::move(M), Mode, IROutputOnly,
-                                      EmitOnlyKernelsAsEntryPoints,
-                                      DeviceGlobals);
-
-  StringRef FileSuffix = IsEsimd ? "esimd_" : "";
-
-  for (size_t I = 0; I < MSplit->totalSplits(); ++I) {
-    module_split::ModuleDesc ResMDesc = MSplit->nextSplit();
-    Module &ResM = ResMDesc.getModule();
-
-    bool SpecConstsMet = processSpecConstants(ResM);
-    bool CompileTimePropertiesMet = processCompileTimeProperties(ResM);
-
-    if (IROutputOnly) {
-      // the result is the transformed input LLVM IR file rather than a file
-      // table
-      saveModuleIR(ResM, OutputFilename);
-      return TblFiles;
-    }
-
-    {
-      // Reuse input module with only regular SYCL kernels if there were
-      // no spec constants and no splitting.
-      // We cannot reuse input module for ESIMD code since it was transformed.
-      std::string ResModuleFile{};
-      bool CanReuseInputModule = !SyclAndEsimdCode && !IsEsimd &&
-                                 !IsLLVMUsedRemoved && !SpecConstsMet &&
-                                 !CompileTimePropertiesMet &&
-                                 (MSplit->totalSplits() == 1);
-      if (CanReuseInputModule)
-        ResModuleFile = InputFilename;
-      else {
-        ResModuleFile =
-            makeResultFileName((OutputAssembly) ? ".ll" : ".bc", I, FileSuffix);
-        saveModuleIR(ResM, ResModuleFile);
+  Modified |= removeSYCLKernelsConstRefArray(*M.get());
+
+  // Do invoke_simd processing before splitting because this:
+  // - saves processing time (the pass is run once, even though on larger IR)
+  // - doing it before SYCL/ESIMD splitting is required for correctness
+  Modified |= lowerInvokeSimd(*M);
```

Comment:
This is a new addition for `invoke_simd` processing. This t-form was dormant in the code base until now.